scylla/src/statement/prepared_statement.rs

```diff
@@ -48,24 +50,62 @@ impl PreparedStatement {
     /// Computes the partition key of the target table from given values
     /// Partition keys have a specific serialization rules.
     /// Ref: https://github.com/scylladb/scylla/blob/40adf38915b6d8f5314c621a94d694d172360833/compound_compat.hh#L33-L47
-    pub fn compute_partition_key(&self, bound_values: &[Value]) -> Bytes {
+    pub fn compute_partition_key(
+        &self,
+        bound_values: &SerializedValues,
+    ) -> Result<Bytes, PartitionKeyError> {
         let mut buf = BytesMut::new();
+
         if self.metadata.pk_indexes.len() == 1 {
-            if let Value::Val(v) = &bound_values[self.metadata.pk_indexes[0] as usize] {
-                buf.extend_from_slice(&v[..]);
+            if let Some(v) = bound_values
+                .iter()
+                .nth(self.metadata.pk_indexes[0] as usize)
+                .ok_or_else(|| {
+                    PartitionKeyError::NoPkIndexValue(
+                        self.metadata.pk_indexes[0],
+                        bound_values.len(),
+                    )
+                })?
+            {
+                buf.extend_from_slice(v);
             }
-            return buf.into();
+            return Ok(buf.into());
         }
         // TODO: consider what happens if a prepared statement is of type (?, something, ?),
         // where all three parameters form a partition key. The middle one is not available
         // in bound values.
-        for &pk_index in &self.metadata.pk_indexes {
-            if let Value::Val(v) = &bound_values[pk_index as usize] {
-                buf.put_u16(v.len() as u16);
-                buf.extend_from_slice(&v[..]);
+
+        // TODO: Optimize - maybe we could check if pk_indexes are sorted and do an allocation-free algorithm then?
```

Comment:
It is already `O(value_count + pk_count)` because all values get collected to a `Vec`.
```rust
let values: Vec<Option<&[u8]>> = bound_values.iter().collect();
//...
for {
    let next_val: &Option<&[u8]> = values.get(pk_index as usize)
}
```
This could be a single allocation (maybe without collect)

My thinking was that usually pk_indexes would be sorted and then we could just do a two pointer sweep. Only unsorted would require collecting indexes/values.
People tend to specify values in the order defined in columns(?)

Reply:
Ah, sorry, I misread. I thought it was going `.iter().nth()` in a loop, for some reason...

Yeah, the idea for the optimization makes sense.